### DISCUSSION TITLE
Welcome to Elixir WebRTC Discussions!

### DISCUSSION THREAD

**Discussion by mickel8** [elixir-webrtc/ex_webrtc#71] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Share your use-cases.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by type1fool** [comment]
Hey @mickel8! This is a very interesting collection of repos, and it has my wheels spinning. 🤔 

## Background

I am using WebRTC data channels for low-latency transfer between clients, and I'm wondering if the `ex_webrtc` package would have a role to play in my application. My approach has been to build a `phx-hook` and use a LiveView as a signaling server. This works quite well.

I'm in the process of de-coupling and refactoring the prototype code. This includes building a WebRTC LiveComponent for my app.

## Questions

After poking around the Elixir WebRTC documentation, I appreciate the structure of the modules. What I don't quite get yet is how I'd use the package.

- Would this be used to establish WebRTC connections between servers?
- Does it support client-server communication - essentially as a WebSocket alternative for some purposes?
- All/none of the above?

Excellent work! Can't wait to see more documentation and development.

**Reply by mickel8** [reply]
Hi @type1fool!

Elixir WebRTC is a W3C WebRTC standard implementation so we offer the same things you can do with the [official JS API](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection).

WebRTC does not include a signaling layer so you have to forward ICE/SDP messages on your own between clients using e.g. a Web Socket.

We don't have support for datachannels yet :( 

So to sum up:
> Would this be used two establish WebRTC connections between servers?

Yes, you can use WebRTC to connect two servers.

> Does it support client-server communication - essentially as a WebSocket alternative for some purposes?

In WebRTC there is basically no distinction between a client and a server. We more talk about peers where on of them can be a server but that's not really important as WebRTC supports NAT traversal. 

Does it help?

**Reply by type1fool** [reply]
Got it. Something just clicked for me.

I've been wondering how I could measure WebRTC performance between regions without spinning up headless Chrome on some distributed nodes. This may be a great use case for `ex_webrtc`. I could spin up some Fly machines and connect them via WebRTC using this package to measure latency and connection quality between regions. Knowing this in advance of actual user sessions would be valuable.

**Reply by type1fool** [reply]
Maybe simple UDP tests could accomplish the same, but I wouldn't be surprised if the performance differs a bit from WebRTC.

**Reply by mickel8** [reply]
You can definitely use ex_webrtc for this! And I would love to see the results! Just keep in mind that Fly.io might have some troubles with WebRTC. At least from our experience it was hard to establish UDP ICE connection with a server running on Fly.io but maybe something has changed :thinking:

**Reply by type1fool** [reply]
ICE negotiation has been working well on my prototype that uses a LiveView as a signaling server. The app is hosted on Fly.

They do have [some documentation](https://fly.io/docs/networking/udp-and-tcp/#main-content-start) for UDP, but there's no mention of ICE negotiation in the docs. Were you doing ICE over LV or something else?

**Reply by mickel8** [reply]
So you are able to establish stable UDP ICE connection on Fly? We were using Web Sockets for signaling. Also, this was a couple of months ago so maybe something had changed

**Reply by type1fool** [reply]
Actually, yeah, I have seen some instability. I assumed I had written some buggy WebRTC code, though. It's also been a couple months now since I was working on and testing that part of the application. 

Once the WebRTC connections are established, the peers are connected directly via UDP, so I am unsure if/how Fly would have any impact. For now, I'm using an external vendor's STUN service and disallowing TURN. Have you deployed the same code to another PaaS with better success?

**Reply by mickel8** [reply]
> Once the WebRTC connections are established, the peers are connected directly via UDP, so I am unsure if/how Fly would have any impact. For now, I'm using an external vendor's STUN service and disallowing TURN. Have you deployed the same code to another PaaS with better success?

Nope, we instead switched to Hetzner where we have a bare machine :slightly_smiling_face:

**Reply by type1fool** [reply]
![interesting](https://github.com/elixir-webrtc/ex_webrtc/assets/13895134/a5ac52df-444b-4920-bf70-3c2d1a2a928e)

**Comment by ambareesha7** [comment]
Hey @mickel8 thank you for WebRTC in Elixir,
context: I run a [YouTube channel](https://www.youtube.com/@try_with_atharva) and trying to build feature like Google meet in LiveView for YouTube tutorial `WebRTC in LiveView` still working on it.

How to disable these logs?
![Screenshot 2024-09-18 at 2 55 41 PM](https://github.com/user-attachments/assets/e9146dd6-e301-4385-abb2-d15bba5aeb90)

**Reply by mickel8** [reply]
Hi @ambareesha7, nice to hear that!

These are debug logs from our ex_ice library. You either have to change your log level from debug to info or disable logs from ExICE modules or ex_ice app.

See [here](https://hexdocs.pm/logger/Logger.html#module-compile-configuration)

**Reply by ambareesha7** [reply]
thank you
